File: frontend/src/lib/types/recheck.ts
import type { ScanRead } from '$lib/types/scan';

export type RecheckTone = 'up' | 'down' | 'neutral';

export interface RecheckChange {
	field: string;
	label: string;
	before: string | null;
	after: string | null;
	tone: RecheckTone;
}

export interface Recheck {
	id: string;
	scan_id: string;
	parent_scan_id: string;
	dimension: string;
	asset_kind: string;
	asset_key: string;
	changed: boolean;
	changes: RecheckChange[];
	created_at: string;
	status: string;
	stage_titles: string[];
	duration_seconds: number | null;
}

export interface RescanDimension {
	dimension: string;
	label: string;
	noun: string;
	noun_plural: string;
	seed_kind: string;
	default_stages: string[];
}

export interface RescanSchema {
	dimensions: RescanDimension[];
	rescannable_stages: string[];
	max_assets: number;
	max_scans: number;
}

export interface SeedPick {
	value: string;
	scan_id?: string;
}

export interface QuerySelection {
	filter: Record<string, unknown>;
	scan_ids: string[];
}

export interface SeedSelection {
	dimension: string;
	picks?: SeedPick[];
	query?: QuerySelection | null;
	exclude?: string[];
}

export interface RescanCreate {
	parent_scan_id?: string | null;
	dimension?: string;
	assets?: string[];
	selection?: SeedSelection | null;
	stages?: string[];
	overrides?: Record<string, Record<string, unknown>>;
	context_id?: string | null;
	intensity?: string | null;
	template_ids?: string[];
}

export interface SeedGroupSummary {
	target_id: string;
	target_value: string;
	scan_id: string;
	count: number;
}

export interface RunPreview {
	dimension: string;
	seed_kind: string;
	asset_count: number;
	matched: number | null;
	target_count: number;
	capped: boolean;
	targets: SeedGroupSummary[];
	stage_titles: string[];
}

export interface FocusedRun {
	run_group_id: string;
	scans: ScanRead[];
	asset_count: number;
	matched: number | null;
	target_count: number;
	capped: boolean;
	stage_titles: string[];
}

File: shared/models/recheck.py
import uuid
from datetime import datetime

from pydantic import BaseModel
from sqlalchemy import Column
from sqlalchemy.types import JSON
from sqlmodel import Field, SQLModel

from shared.utils.datetime import utc_now


class AssetRecheck(SQLModel, table=True):
    __tablename__ = "asset_rechecks"

    id: uuid.UUID = Field(default_factory=uuid.uuid4, primary_key=True, index=True)
    project_id: uuid.UUID = Field(foreign_key="projects.id", index=True)
    target_id: uuid.UUID = Field(
        foreign_key="targets.id", index=True, ondelete="CASCADE"
    )
    scan_id: uuid.UUID = Field(foreign_key="scans.id", index=True, ondelete="CASCADE")
    parent_scan_id: uuid.UUID = Field(
        foreign_key="scans.id", index=True, ondelete="CASCADE"
    )
    dimension: str = Field(max_length=32)
    asset_kind: str = Field(max_length=16)
    asset_key: str = Field(max_length=500, index=True)
    changed: bool = Field(default=False)
    changes: list = Field(default_factory=list, sa_column=Column(JSON, nullable=False))
    created_at: datetime = Field(default_factory=utc_now)


class RecheckChange(BaseModel):
    field: str
    label: str
    before: str | None = None
    after: str | None = None
    tone: str = "neutral"


class RecheckRead(BaseModel):
    id: uuid.UUID
    scan_id: uuid.UUID
    parent_scan_id: uuid.UUID
    dimension: str
    asset_kind: str
    asset_key: str
    changed: bool
    changes: list[RecheckChange]
    created_at: datetime
    status: str
    stage_titles: list[str] = []
    duration_seconds: float | None = None
